near-plugins-derive/src/access_controllable.rs

```diff
@@ -0,0 +1,287 @@
+        #[near_bindgen]
+        impl AccessControllable for #ident {
+            fn acl_storage_prefix() -> &'static [u8] {
+                (#storage_prefix).as_bytes()
+            }
+
+            #[private]
+            fn acl_grant_role_unchecked(&mut self, role: String, account_id: ::near_sdk::AccountId) -> bool {
+                let role = <#role_type>::try_from(role.as_str()).expect(#ERR_PARSE_ROLE);
+                self.#acl_field.grant_role_unchecked(role, &account_id)
+            }
+
+            fn acl_has_role(&self, role: String, account_id: ::near_sdk::AccountId) -> bool {
+                let role = <#role_type>::try_from(role.as_str()).expect(#ERR_PARSE_ROLE);
+                self.#acl_field.has_role(role, &account_id)
+            }
+
+            fn acl_has_any_role(&self, roles: Vec<String>, account_id: ::near_sdk::AccountId) -> bool {
+                let roles: Vec<#role_type> = roles
+                    .iter()
+                    .map(|role| <#role_type>::try_from(role.as_str()).expect(#ERR_PARSE_ROLE))
+                    .collect();
+                self.#acl_field.has_any_role(roles, &account_id)
+            }
+        }
+    };
+
+    output.into()
+}
+
+fn inject_acl_field(
+    item: &mut ItemStruct,
+    field_name: &syn::Ident,
+    acl_type: &syn::Ident,
+) -> Result<(), syn::Error> {
+    let fields = match item.fields {
+        syn::Fields::Named(ref mut fields) => fields,
+        _ => {
+            return Err(syn::Error::new(
+                item.ident.span(),
+                "Expected to have named fields",
+            ))
+        }
+    };
+
+    fields.named.push(syn::Field::parse_named.parse2(quote! {
+        #field_name: #acl_type
+    })?);
+    Ok(())
+}
+
+#[derive(Debug, FromMeta)]
+pub struct MacroArgsAny {
+    roles: darling::util::PathList,
+}
+
+pub fn access_control_any(attrs: TokenStream, item: TokenStream) -> TokenStream {
+    let attr_args = parse_macro_input!(attrs as AttributeArgs);
+    let cloned_item = item.clone();
+    let input: ItemFn = parse_macro_input!(cloned_item);
+    if is_near_bindgen_wrapped_or_marshall(&input) {
+        return item;
+    }
+
+    let ItemFn {
+        attrs,
+        vis,
+        sig,
+        block,
+    } = input;
+    let function_name = sig.ident.to_string();
+    let stmts = &block.stmts;
+
+    let macro_args = match MacroArgsAny::from_list(&attr_args) {
+        Ok(args) => args,
+        Err(e) => {
+            return TokenStream::from(e.write_errors());
+        }
+    };
+    let roles = macro_args.roles;
+    assert!(roles.len() > 0, "Specify at least one role");
+
+    let acl_check = quote! {
+        let __roles: Vec<&str> = vec![#(#roles.into()),*];
+        let __roles_ser: Vec<String> = __roles.iter().map(|&role| role.into()).collect();
+        let __account_id = ::near_sdk::env::predecessor_account_id();
+        if !self.acl_has_any_role(__roles_ser, __account_id) {
```

Comment:
Add TODO for a future revision that handles the case where a single role is passed so that it is treated simpler (no vector manipulation, and call directly to `acl_has_role`). I consider it relevant because I expect this to be the most common case.

Reply:
Done in [f60faa8](https://github.com/aurora-is-near/near-plugins/pull/5/commits/f60faa824d5bb134bded8d9073dd371d065aad5c).